src/x2attr.c

```diff
@@ -0,0 +1,649 @@
+			_("Attribute definition error at line %ld in file \"%s\": "
+			   "File attribute %s already defined"),
+			linenum, filename, k);
+        return;
+    }
+    
+    // Good to go...  Casts needed because of glib.
+    g_hash_table_insert(d->hashtable, (void *)k, (void *)val);
+    
+    dprintf("Inserted attr %s=%s in table %c\n",
+            k, val, (char)type);
+}
+
+void 
+_x2attr_destroy_dicts(struct gerb_state * gs)
+{
+    _destroy_dict(gs->file_attrs);
+    _destroy_dict(gs->aperture_attrs);
+    _destroy_dict(gs->object_attrs);
+}
+
+x2attr_dict_t * 
+_x2attr_new_dict()
+{
+    return _new_dict(X2ATTR_ANY);
+}
+
+
+const char * 
+_x2attr_value_changed(x2attr_dict_t * d, const char * key, const char * value)
+{
+    const char * oldval = (const char *)g_hash_table_lookup(d->hashtable, key);
+    if (oldval != value)
+        g_hash_table_insert(d->hashtable, (void *)key, (void *)value);
+    return oldval == value ? NULL : value;
+}
+
+void 
+_x2attr_destroy_dict(x2attr_dict_t * d)
+{
+    _destroy_dict(d);
+}
+
+
+
+/******************
+  Library API
+  
+  Manipulates attributes programmatically i.e. not when reading
+  RS274-X2.
+  
+  Operate on
+    image->attrs
+    aperture->attrs
+    net->attrs.
+    
+*******************/
+
+
+void 
+x2attr_foreach_image_attr(gerbv_image_t * image, x2attr_enumeration_callback_t callback, gpointer user_data)
+{
+    _enumerate(image->attrs, callback, user_data);
+}
+
+void 
+x2attr_foreach_aperture_attr(gerbv_aperture_t * aperture, x2attr_enumeration_callback_t callback, gpointer user_data)
+{
+    _enumerate(aperture->attrs, callback, user_data);
+}
+
+void 
+x2attr_foreach_net_attr(gerbv_net_t * net, x2attr_enumeration_callback_t callback, gpointer user_data)
+{
+    _enumerate(net->attrs, callback, user_data);
+}
+
+
+
+const char * 
+x2attr_get_image_attr(const gerbv_image_t * image, const char * key)
+{
+    return _find_value(image->attrs, g_intern_string(key));
+}
+
+const char * 
+x2attr_get_image_attr_or_default(const gerbv_image_t * image, 
+                const char * key, const char * dflt)
+{
+    const char * v = _find_value(image->attrs, g_intern_string(key));
+    return v ? v : dflt;
+}
+
+const char * 
+x2attr_get_aperture_attr(const gerbv_aperture_t * aperture, const char * key)
+{
+    return _find_value(aperture->attrs, g_intern_string(key));
+}
+
+const char * 
+x2attr_get_aperture_attr_or_default(const gerbv_aperture_t * aperture, 
+                const char * key, const char * dflt)
+{
+    const char * v = _find_value(aperture->attrs, g_intern_string(key));
+    return v ? v : dflt;
+}
+
+const char * 
+x2attr_get_net_attr(const gerbv_net_t * net, const char * key)
+{
+    return _find_chained_value(net->attrs, g_intern_string(key));
+}
+
+const char * 
+x2attr_get_net_attr_or_default(const gerbv_net_t * net, 
+                const char * key, const char * dflt)
+{
+    const char * v = _find_chained_value(net->attrs, g_intern_string(key));
```

Comment:
same as above.

How come this one is chained and the rest of them aren't?  Why don't we always use a chained version?

Reply:
The chaining is only used in this case, according to the standard, i.e. for "net" attributes.  In some cases, after looking for a net attribute, also look for aperture attributes.  IIRC, this applies to polygon regions.